pkg/services/ngalert/prom/convert.go

```diff
@@ -211,10 +264,84 @@ func createAlertQueryNode(datasourceUID, datasourceType, expr string, fromTimeRa
 	return models.AlertQuery{
 		DatasourceUID: datasourceUID,
 		Model:         modelJSON,
-		RefID:         "A",
+		RefID:         queryRefID,
 		RelativeTimeRange: models.RelativeTimeRange{
 			From: models.Duration(fromTimeRange + evaluationOffset),
 			To:   models.Duration(0 + evaluationOffset),
 		},
 	}, nil
 }
+
+func createMathNode() (models.AlertQuery, error) {
+	modelData := map[string]interface{}{
+		"datasource": map[string]interface{}{
+			"name": expr.TypeCMDNode.String(),
+			"type": expr.DatasourceType,
+			"uid":  expr.DatasourceUID,
+		},
+		"conditions": []map[string]interface{}{
+			{
+				"type":     "query",
+				"reducer":  map[string]interface{}{"type": "avg", "params": []interface{}{}},
+				"operator": map[string]interface{}{"type": "and"},
+				"query":    map[string]interface{}{"params": []interface{}{}},
+				"evaluator": map[string]interface{}{
+					"type":   "gt",
+					"params": []interface{}{0, 0},
+				},
+			},
+		},
+		"refId":      prometheusMathRefID,
+		"type":       string(expr.QueryTypeMath),
+		"expression": fmt.Sprintf("is_number($%[1]s) || is_nan($%[1]s) || is_inf($%[1]s)", queryRefID),
+	}
+
+	modelJSON, err := json.Marshal(modelData)
+	if err != nil {
+		return models.AlertQuery{}, err
+	}
+
+	return models.AlertQuery{
+		DatasourceUID: expr.DatasourceUID,
+		Model:         modelJSON,
+		RefID:         prometheusMathRefID,
+		QueryType:     string(expr.QueryTypeMath),
+	}, nil
+}
+
+func createThresholdNode() (models.AlertQuery, error) {
+	modelData := map[string]interface{}{
+		"datasource": map[string]interface{}{
+			"name": expr.TypeCMDNode.String(),
+			"type": expr.DatasourceType,
+			"uid":  expr.DatasourceUID,
+		},
+		"conditions": []map[string]interface{}{
+			{
+				"type":     "query",
+				"reducer":  map[string]interface{}{"type": "last", "params": []interface{}{}},
+				"operator": map[string]interface{}{"type": "and"},
+				"query":    map[string]interface{}{"params": []interface{}{}},
+				"evaluator": map[string]interface{}{
+					"type":   string(expr.ThresholdIsAbove),
+					"params": []interface{}{0},
+				},
+			},
+		},
+		"refId":      thresholdRefID,
+		"type":       string(expr.QueryTypeThreshold),
+		"expression": prometheusMathRefID,
+	}
```

Comment:
Let's move it to `expr` package and create a factory functions for these expressions that will validate the input (in case of math expr)

Reply:
I added a test for the math expression to ensure it can be parsed by the `mathexpr` package without errors. I also moved the queries to a separate file for now, using structures from `expr`.